### PR TITLE
Use std_{code,expr,type} constructors

### PR DESCRIPTION
**Pull request by tautschnig**
Applies resource-acquisition-is-initialisation. This might slightly improve
performance, but otherwise should not change behaviour. While it will ensure
consistent data structures, no bugs were spotted/fixed here.

### CODE REVIEW

File: src/goto-instrument/accelerate/acceleration_utils.cpp
Comment (kroening) [inline src/goto-instrument/accelerate/acceleration_utils.cpp]: forall is always boolean, so I'd be tempted to tweak the class definition of forall_exprt to derive from binary_predicate_exprt instead of binary_exprt.

Comment (kroening) [reply src/goto-instrument/accelerate/acceleration_utils.cpp]: Oh, and also to add a constructor that takes two arguments.

File: src/goto-instrument/accelerate/acceleration_utils.cpp
Comment (kroening) [inline src/goto-instrument/accelerate/acceleration_utils.cpp]: same here

File: src/goto-instrument/accelerate/acceleration_utils.cpp
Comment (kroening) [inline src/goto-instrument/accelerate/acceleration_utils.cpp]: and again

File: src/goto-instrument/accelerate/acceleration_utils.cpp
Comment (kroening) [inline src/goto-instrument/accelerate/acceleration_utils.cpp]: I would make that one const, which makes it clearer that 'k' later on doesn't change

File: src/goto-instrument/accelerate/polynomial_accelerator.cpp
Comment (kroening) [inline src/goto-instrument/accelerate/polynomial_accelerator.cpp]: another one

File: src/solvers/qbf/qbf_bdd_core.cpp
Comment (kroening) [inline src/solvers/qbf/qbf_bdd_core.cpp]: Simply use integer_typet()?

File: src/solvers/qbf/qbf_squolem_core.cpp
Comment (kroening) [inline src/solvers/qbf/qbf_squolem_core.cpp]: Same here.

**Review by kroening — APPROVED** [review]
Only minor tweaks, the forall tweaks could be done separately.

File: src/analyses/invariant_propagation.cpp
Comment (thk123) [inline src/analyses/invariant_propagation.cpp]: No need to do it as part of this PR, but seems like would make more sense to introduce this _inside_ the for loop then you can construct from the `component` directly.

Comment (thk123) [reply src/analyses/invariant_propagation.cpp]: But you should probably use `set_component_name`  on line 108 here

File: src/analyses/invariant_set.cpp
Comment (thk123) [inline src/analyses/invariant_set.cpp]: Use `set_component_name` on the lines below

File: src/ansi-c/c_typecheck_expr.cpp
Comment (thk123) [inline src/ansi-c/c_typecheck_expr.cpp]: `dereference_exprt deref{expr.op0()}`?

Comment (thk123) [reply src/ansi-c/c_typecheck_expr.cpp]: Or perhaps since this doesn't actually set the type, perhaps `nil_typet` or something to precisely mimic existing behaviour?

File: src/cpp/cpp_typecheck_constructor.cpp
Comment (thk123) [inline src/cpp/cpp_typecheck_constructor.cpp]: Nit annotated the TODO with your name `TODO(mtautschnig): `

File: src/cpp/cpp_typecheck_constructor.cpp
Comment (thk123) [inline src/cpp/cpp_typecheck_constructor.cpp]: Nit annotated the TODO with your name `// TODO(mtautschnig):`

File: src/cpp/cpp_typecheck_destructor.cpp
Comment (thk123) [inline src/cpp/cpp_typecheck_destructor.cpp]: Should be able to use the  constructor for setting `op` here?

**Review by thk123 — COMMENTED** [review]
Got to `src/goto-instrument/accelerate/polynomial_accelerator.cpp` if someone wants to pick up from there, but looks like a nice clean up!

**Comment by tautschnig** [comment]
@thk123 @kroening Thanks a lot for all the comments, all addressed.

**Comment by tautschnig** [comment]
I note that this should go in after #1897 (and it will require a conflict-resolving rebase at that time).

**Review by martin-cs — APPROVED** [review]
I agree with the general idea; irept's becoming more functional is not a bad thing.  I'm concerned that this kind of wide change, even if localised, will break external patch sets.  Hearing from a few people that it wouldn't would help me rest more easily in that regard.

**Comment by smowton** [comment]
Don't really have time to read this in detail today, sorry, so I'm going to duck in favour of another code owner

File: src/analyses/goto_check.cpp
Comment (romainbrenguier) [inline src/analyses/goto_check.cpp]: const

File: src/analyses/goto_check.cpp
Comment (romainbrenguier) [inline src/analyses/goto_check.cpp]: const

File: src/analyses/goto_check.cpp
Comment (romainbrenguier) [inline src/analyses/goto_check.cpp]: const here and on all the other modified lines of this file

File: src/ansi-c/ansi_c_entry_point.cpp
Comment (romainbrenguier) [inline src/ansi-c/ansi_c_entry_point.cpp]: const ? (applies to the other changes in this file)

File: src/cbmc/bv_cbmc.cpp
Comment (romainbrenguier) [inline src/cbmc/bv_cbmc.cpp]: You can remove the blank lines and add const everywhere

File: src/goto-instrument/accelerate/acceleration_utils.cpp
Comment (romainbrenguier) [inline src/goto-instrument/accelerate/acceleration_utils.cpp:1064]: can it be a reference?

File: src/goto-instrument/accelerate/acceleration_utils.cpp
Comment (romainbrenguier) [inline src/goto-instrument/accelerate/acceleration_utils.cpp]: `const and_exprt k_bound(`

File: src/solvers/floatbv/float_bv.cpp
Comment (romainbrenguier) [inline src/solvers/floatbv/float_bv.cpp]: Could use `disjunction` instead

File: src/goto-instrument/accelerate/acceleration_utils.cpp
Comment (tautschnig) [reply src/goto-instrument/accelerate/acceleration_utils.cpp:1064]: No, `symbolt::symbol_expr` returns by value.

**Review by thk123 — APPROVED** [review]
Remembered only need to review `analyses` and `java_bytecode` - these look fine (and previous suggestions have been applied) so happy for this to go in

**Comment by tautschnig** [comment]
I've just found another 250 locations that may be amenable to this sort of improvement, I'll post updates later on today.

**Comment by martin-cs** [comment]
On Fri, 2018-03-02 at 10:48 +0000, Michael Tautschnig wrote:
> I've just found another 250 locations that may be amenable to this sort of improvement, I'll post updates later on today.

<expression of concern about the size of the resultant change and the
impact on other patches>

**Comment by tautschnig** [comment]
> expression of concern about the size of the resultant change and the impact on other patches

Concern noted and shared; but also those changes seem to be the right-thing-to-do(TM). Does it help in any way that the code patched here typically hasn't been touched in a long time as otherwise it wouldn't have been written in this style in the first place?

**Comment by kroening** [comment]
I am positive about these sort of changes; they clearly improve readability, maintainability and robustness. Will merge once CIed.

**Pull request merged**
